File: src/test/java/com/natu/ftax/transaction/domain/TransactionTest.java
package com.natu.ftax.transaction.domain;

import com.natu.ftax.common.exception.FunctionalException;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class TransactionTest {

    @Test
    void shouldCreateDraftTransaction() {
        Transaction draftTransaction = Transaction.create("123");

        assertEquals("123", draftTransaction.getId());
        assertEquals("draft", draftTransaction.getState());
    }

    @Test
    void shouldCreateTransactionFromDraftTransaction() {
        Transaction transaction = Transaction.create("abc");

        transaction.submit(
                new SubmitTransactionCommand(
                        "abc",
                        "swap",
                        Instant.now(),
                        "0x123",
                        "0x456",
                        "0x789",
                        new BigDecimal("120.0"),
                        BigDecimal.ZERO,
                        BigDecimal.ZERO,
                        "0x124"
                )
        );

        assertEquals("abc", transaction.getId());
        assertEquals("submitted", transaction.getState());
        assertEquals("swap", transaction.getTransactionType());
        assertEquals("0x123", transaction.getTokenIn().getSymbol());
        assertEquals("0x456", transaction.getTokenOut().getSymbol());
        assertEquals("0x789", transaction.getTokenFee().getSymbol());
        assertEquals(new BigDecimal("120.0"), transaction.getAmountIn());
        assertEquals(BigDecimal.ZERO, transaction.getAmountOut());
        assertEquals(BigDecimal.ZERO, transaction.getAmountFee());
        assertEquals("0x124", transaction.getExternalId());
    }

    @ParameterizedTest
    @MethodSource("amountArguments")
    void shouldNotBeAbleToSubmitTransactionWithNegativeAmounts(
            BigDecimal amountIn,
            BigDecimal amountOut, BigDecimal amountFee) {
        Transaction tx = Transaction.create("anId");

        Executable executable = () -> tx.submit(
                new SubmitTransactionCommand(
                        "anId",
                        "swap",
                        Instant.now(),
                        "0x123",
                        "0x456",
                        "0x789",
                        amountIn,
                        amountOut,
                        amountFee,
                        "0x123"
                )
        );

        FunctionalException exception = assertThrows(FunctionalException.class, executable);
        Assertions.assertThat(exception.getMessage()).contains("Value cannot be negative");
    }

    @Test
    void shouldBeAbleToEditTransaction() {
        Transaction transaction = submittedTransaction();

        transaction.edit();

        assertEquals("abc", transaction.getId());
        assertEquals("draft", transaction.getState());
    }

    @Test
    void shouldNotBeAbleToEditDraftTransaction() {
        Transaction transaction = Transaction.create("abc");

        Executable executable = transaction::edit;
        FunctionalException exception = assertThrows(FunctionalException.class, executable);

        Assertions.assertThat(exception.getMessage()).contains("Transaction is not in SUBMITTED state");
    }

    private Transaction submittedTransaction() {
        Transaction transaction = Transaction.create("abc");

        transaction.submit(
                new SubmitTransactionCommand(
                        "abc",
                        "swap",
                        Instant.now(),
                        "0x123",
                        "0x456",
                        "0x789",
                        BigDecimal.valueOf(120.0),
                        BigDecimal.ZERO,
                        BigDecimal.ZERO,
                        "0x124"
                )
        );
        return transaction;
    }

    private static Stream<Arguments> amountArguments() {
        return Stream.of(
                Arguments.of(BigDecimal.valueOf(-1.0), BigDecimal.ZERO, BigDecimal.ZERO),
                Arguments.of(BigDecimal.ZERO, BigDecimal.valueOf(-1.0), BigDecimal.ZERO),
                Arguments.of(BigDecimal.ZERO, BigDecimal.ZERO, BigDecimal.valueOf(-1.0))
        );
    }
}

File: src/test/kotlin/com/natu/ftax/transaction/domain/TransactionTest.kt
package com.natu.ftax.transaction.domain

import com.natu.ftax.common.exception.FunctionalException
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import java.util.*
import java.util.stream.Stream

class TransactionTest {

    @Test
    fun `should create a draft transaction`() {
        val draftTransaction = Transaction.create("123")

        assertEquals("123", draftTransaction.id)
        assertEquals(TransactionState.DRAFT, draftTransaction.state)
    }

    @Test
    fun `should create a transaction from draft transaction`() {
        val transaction = Transaction.create("abc")

        transaction.submit(
            SubmitTransactionCommand(
                "abc",
                TransactionType.SWAP,
                Date(),
                "0x123",
                "0x456",
                "0x789",
                120.0,
                0.0,
                0.0,
                "0x124"
            )
        )

        assertEquals("abc", transaction.id)
        assertEquals(TransactionState.SUBMITTED, transaction.state)
        assertEquals(TransactionType.SWAP, transaction.transactionType)
        assertEquals("0x123", transaction.token1)
        assertEquals("0x456", transaction.token2)
        assertEquals("0x789", transaction.tokenFee)
        assertEquals(120.0, transaction.amount1)
        assertEquals(0.0, transaction.amount2)
        assertEquals(0.0, transaction.amountFee)
        assertEquals("0x124", transaction.externalId)

    }

    @ParameterizedTest
    @MethodSource("amountArguments")
    fun `should not be able to submit a transaction with negative amounts`(
        amount1: Double,
        amount2: Double,
        amountFee: Double
    ) {
        val tx = Transaction.create("anId")
        val exception = assertThrows<IllegalArgumentException> {
            tx.submit(
                SubmitTransactionCommand(
                    "anId",
                    TransactionType.SWAP,
                    Date(),
                    "0x123",
                    "0x456",
                    "0x789",
                    amount1,
                    amount2,
                    amountFee,
                    "0x123"
                )
            )
        }



        Assertions.assertThat(exception.message).contains(" must not be less than 0")
    }

    @Test
    fun `should be able to edit a transaction`() {
        val transaction = submittedTransaction()

        transaction.edit()

        assertEquals("abc", transaction.id)
        assertEquals(TransactionState.DRAFT, transaction.state)
    }

    @Test
    fun `should not be able to edit a draft transaction`() {
        val transaction = Transaction.create("abc")

        val exception = assertThrows<FunctionalException> {
            transaction.edit()
        }

        Assertions.assertThat(exception.message).contains("Transaction is not in SUBMITTED state")
    }

    private fun submittedTransaction(): Transaction {
        val transaction = Transaction.create("abc")

        transaction.submit(
            SubmitTransactionCommand(
                "abc",
                TransactionType.SWAP,
                Date(),
                "0x123",
                "0x456",
                "0x789",
                120.0,
                0.0,
                0.0,
                "0x124"
            )
        )
        return transaction
    }


    companion object {
        @JvmStatic
        fun amountArguments(): Stream<Arguments> = Stream.of(
            Arguments.of(-1.0, 0.0, 0.0),
            Arguments.of(0.0, -1.0, 0.0),
            Arguments.of(0.0, 0.0, -1.0)
        )
    }
}